### ISSUE TITLE
Status of pods can become "OutOfCpu" when many pods are created and completed in a short time on the same node.

### ISSUE THREAD

**Issue by shioshiota** [kubernetes/kubernetes#106884]
### What happened?

Status of pods can become "OutOfCpu" when many pods are created and completed in a short time on the same node.

### What did you expect to happen?

Status of pods should become "Running", "Pending" or "Completed".

### How can we reproduce it (as minimally and precisely as possible)?

I created a procedure to reproduce this state.
https://gist.github.com/shioshiota/84f1f4e162c335fe551c60a118676c3c

### Anything else we need to know?

We couldn't reproduce this behavior in kubernetes 1.21.2 cluster.
We found this issue in kubernetes 1.22.4.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T21:04:39Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.4", GitCommit:"b695d79d4f967c403a96986f1750a35eb75e75f1", GitTreeState:"clean", BuildDate:"2021-11-23T01:42:25Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"linux/arm64"}
```

</details>

### Cloud provider

<details>
no
</details>

### OS version

<details>

```console
# On Linux:
$ docker exec -it kind-control-plane cat /etc/os-release
PRETTY_NAME="Ubuntu 21.10"
NAME="Ubuntu"
VERSION_ID="21.10"
VERSION="21.10 (Impish Indri)"
VERSION_CODENAME=impish
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=impish
$ docker exec -it kind-control-plane uname -a
Linux kind-control-plane 5.10.76-linuxkit #1 SMP PREEMPT Mon Nov 8 11:22:26 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by pacoxu** [comment]
https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodename

> If the named node does not have the resources to accommodate the pod, the pod will fail and its reason will indicate why, for example, OutOfmemory or OutOfcpu.


There is a known issue  #106419 https://github.com/kubernetes/kubernetes/issues/106419#issuecomment-982952646 that OutOfCpu maybe caused by data race of `kubelet starts a static pod` and `scheduler schedule a pod to the node`.

**Comment by shioshiota** [comment]
We confirmed this situation occurred even if it is in a node that is one without static pods like a worker node too.
I will write a document to reproduce in a kind cluster and share it here.

**Comment by shioshiota** [comment]
@pacoxu 

I reproduced this issue on a node without static pods.
https://gist.github.com/shioshiota/1f9c6e26b6277522bfe563de3af59640

Thanks,

**Comment by ehashman** [comment]
This was fixed: https://github.com/kubernetes/kubernetes/issues/105523

But I don't think it was backported to 1.22 from the looks of things. We confirmed it affected 1.23 and 1.22, fixed in 1.23.

/assign
/triage accepted

**Comment by shioshiota** [comment]
Thank you!
However, I think the patch was backported to 1.22 in this PR https://github.com/kubernetes/kubernetes/pull/105560.
And it was released as 1.22.3
https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#bug-or-regression-1

And I confirmed this issue occurred in 1.22.4.

**Comment by shioshiota** [comment]
I found a log that may be related to this issue.
I looked into kublet log when OutOfCpu errors occurred.

This log may mean a not thread-safe function "updateStatusInternal" calls concurrently.
It shows pod status was updated to version-3 four times.
https://github.com/kubernetes/kubernetes/blob/v1.22.4/pkg/kubelet/status/status_manager.go#L390-L393
```
grep 9093293a-8fc4-40d6-baa2-11ad6dbfa9ad worker.log | grep updateStatusInternal
Dec 09 13:19:52 kind-worker kubelet[172]: I1209 13:19:52.959175     172 status_manager.go:466] "updateStatusInternal" version=1 pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containers="(ubuntu state=waiting previous=<none>)"
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.597273     172 status_manager.go:466] "updateStatusInternal" version=2 pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containers="(ubuntu state=terminated=0 previous=<none>)"
Dec 09 13:20:05 kind-worker kubelet[172]: I1209 13:20:05.603275     172 status_manager.go:466] "updateStatusInternal" version=3 pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containers="(ubuntu state=terminated=0 previous=<none>)"
Dec 09 13:20:06 kind-worker kubelet[172]: I1209 13:20:06.607423     172 status_manager.go:466] "updateStatusInternal" version=3 pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containers="(ubuntu state=terminated=0 previous=<none>)"
Dec 09 13:20:07 kind-worker kubelet[172]: I1209 13:20:07.611700     172 status_manager.go:466] "updateStatusInternal" version=3 pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containers="(ubuntu state=terminated=0 previous=<none>)"
Dec 09 13:20:07 kind-worker kubelet[172]: I1209 13:20:07.613852     172 status_manager.go:360] "TerminatePod calling updateStatusInternal" pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad
Dec 09 13:20:07 kind-worker kubelet[172]: I1209 13:20:07.613871     172 status_manager.go:466] "updateStatusInternal" version=3 pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containers="(ubuntu state=terminated=0 previous=<none>)"
```

**Comment by shioshiota** [comment]
I believe that this log shows the race condition between `PodWorkers.podSyncStatuses` and k8s API server if all containers of a Pod are finished while kubelet processes the Pod event(updateType=0(SyncPodWork)).

```
# PodWorkers started handling event
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.230974     172 pod_workers.go:775] "Processing pod event" pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad updateType=0

# The Pod completed
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.596091     172 generic.go:155] "GenericPLEG" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containerID="61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6" oldState=non-existent newState=exited
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.597165     172 generic.go:296] "Generic (PLEG): container finished" podID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containerID="61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6" exitCode=0
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.597178     172 kubelet.go:2091] "SyncLoop (PLEG): event for pod" pod="default/out-of-cpu-test-fdlpx" event=&{ID:9093293a-8fc4-40d6-baa2-11ad6dbfa9ad Type:ContainerDied Data:61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6}
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.597187     172 kubelet.go:1531] "syncPod enter" pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad

# kubelet reported the Pod was `Completed` without calling `PodWorkers.acknowledgeTerminating()`.
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.597273     172 status_manager.go:466] "updateStatusInternal" version=2 pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad containers="(ubuntu state=terminated=0 previous=<none>)"
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.597318     172 status_manager.go:486] "Status Manager: adding pod with new status to podStatusChannel" pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad statusVersion=2 status={Phase:Succeeded Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-12-09 13:19:52 +0000 UTC Reason:PodCompleted Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-12-09 13:19:52 +0000 UTC Reason:PodCompleted Message:} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-12-09 13:19:52 +0000 UTC Reason:PodCompleted Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-12-09 13:19:52 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:172.18.0.2 PodIP:10.244.1.2 PodIPs:[{IP:10.244.1.2}] StartTime:2021-12-09 13:19:52 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:ubuntu State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:0,Signal:0,Reason:Completed,Message:,StartedAt:2021-12-09 13:20:04 +0000 UTC,FinishedAt:2021-12-09 13:20:04 +0000 UTC,ContainerID:containerd://61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:docker.io/library/ubuntu:18.04 ImageID:docker.io/library/ubuntu@sha256:0fedbd5bd9fb72089c7bbca476949e10593cebed9b1fb9edf5b79dbbacddd7d6 ContainerID:containerd://61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6 Started:0x4000efeaac}] QOSClass:Burstable EphemeralContainerStatuses:[]}
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.597554     172 status_manager.go:166] "Status Manager: syncing pod with status from podStatusChannel" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad statusVersion=2 status={Phase:Succeeded Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-12-09 13:19:52 +0000 UTC Reason:PodCompleted Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-12-09 13:19:52 +0000 UTC Reason:PodCompleted Message:} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-12-09 13:19:52 +0000 UTC Reason:PodCompleted Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-12-09 13:19:52 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:172.18.0.2 PodIP:10.244.1.2 PodIPs:[{IP:10.244.1.2}] StartTime:2021-12-09 13:19:52 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:ubuntu State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:0,Signal:0,Reason:Completed,Message:,StartedAt:2021-12-09 13:20:04 +0000 UTC,FinishedAt:2021-12-09 13:20:04 +0000 UTC,ContainerID:containerd://61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:docker.io/library/ubuntu:18.04 ImageID:docker.io/library/ubuntu@sha256:0fedbd5bd9fb72089c7bbca476949e10593cebed9b1fb9edf5b79dbbacddd7d6 ContainerID:containerd://61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6 Started:0x4000efeaac}] QOSClass:Burstable EphemeralContainerStatuses:[]}
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.603319     172 status_manager.go:621] "Patch status for pod" pod="default/out-of-cpu-test-fdlpx" patch="{\"metadata\":{\"uid\":\"9093293a-8fc4-40d6-baa2-11ad6dbfa9ad\"},\"status\":{\"$setElementOrder/conditions\":[{\"type\":\"Initialized\"},{\"type\":\"Ready\"},{\"type\":\"ContainersReady\"},{\"type\":\"PodScheduled\"}],\"conditions\":[{\"reason\":\"PodCompleted\",\"type\":\"Initialized\"},{\"message\":null,\"reason\":\"PodCompleted\",\"type\":\"Ready\"},{\"message\":null,\"reason\":\"PodCompleted\",\"type\":\"ContainersReady\"}],\"containerStatuses\":[{\"containerID\":\"containerd://61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6\",\"image\":\"docker.io/library/ubuntu:18.04\",\"imageID\":\"docker.io/library/ubuntu@sha256:0fedbd5bd9fb72089c7bbca476949e10593cebed9b1fb9edf5b79dbbacddd7d6\",\"lastState\":{},\"name\":\"ubuntu\",\"ready\":false,\"restartCount\":0,\"started\":false,\"state\":{\"terminated\":{\"containerID\":\"containerd://61fa6c69deee81a155b88c7f9878921aeba0b5d94485451367712f22b5902fe6\",\"exitCode\":0,\"finishedAt\":\"2021-12-09T13:20:04Z\",\"reason\":\"Completed\",\"startedAt\":\"2021-12-09T13:20:04Z\"}}}],\"phase\":\"Succeeded\",\"podIP\":\"10.244.1.2\",\"podIPs\":[{\"ip\":\"10.244.1.2\"}]}}"
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.663649     172 kubelet.go:1744] "syncPod exit" pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad
Dec 09 13:20:04 kind-worker kubelet[172]: I1209 13:20:04.663680     172 pod_workers.go:866] "Processing pod event done" pod="default/out-of-cpu-test-fdlpx" podUID=9093293a-8fc4-40d6-baa2-11ad6dbfa9ad updateType=0
```

kubelet reported the Pod was `Completed` without calling `PodWorkers.acknowledgeTerminating()` since this event was updateType=0(`SyncPodWork`).
https://github.com/kubernetes/kubernetes/blob/b695d79d4f967c403a96986f1750a35eb75e75f1/pkg/kubelet/pod_workers.go#L807-L827

kubernetes scheduler scheduled a new Pod without considering the resources requested by the `Completed` Pod since it scheduled was based on k8s API server state.
As a result, a new Pod might become `OutOfCpu` state since `Kubelet.filterOutTerminatedPods()` didn't filter the `Completed` Pod when it was called by `Kubelet.HandlePodAdditions`.
https://github.com/kubernetes/kubernetes/blob/b695d79d4f967c403a96986f1750a35eb75e75f1/pkg/kubelet/kubelet.go#L2217-L2227
https://github.com/kubernetes/kubernetes/blob/b695d79d4f967c403a96986f1750a35eb75e75f1/pkg/kubelet/kubelet_pods.go#L975-L980

**Comment by shioshiota** [comment]
I patched this and I confirmed this issue was resolved.

```
> git rev-parse HEAD
b695d79d4f967c403a96986f1750a35eb75e75f1

>  git diff
diff --git a/pkg/kubelet/kubelet_pods.go b/pkg/kubelet/kubelet_pods.go
index 6faeb59082e..637cf2b8bca 100644
--- a/pkg/kubelet/kubelet_pods.go
+++ b/pkg/kubelet/kubelet_pods.go
@@ -978,6 +978,12 @@ func (kl *Kubelet) filterOutTerminatedPods(pods []*v1.Pod) []*v1.Pod {
                if kl.podWorkers.IsPodKnownTerminated(p.UID) {
                        continue
                }
+               if st, ok := kl.statusManager.GetPodStatus(p.UID); ok {
+                       if st.Phase == v1.PodSucceeded || st.Phase == v1.PodFailed {
+                               klog.V(4).InfoS("filter terminated pods on status manager", "pod", klog.KOb
+                               continue
+                       }
+               }
                // terminal pods are considered inactive UNLESS they are actively terminating
                if kl.isAdmittedPodTerminal(p) && !kl.podWorkers.IsPodTerminationRequested(p.UID) {
                        continue
```
https://github.com/kubernetes/kubernetes/blob/b695d79d4f967c403a96986f1750a35eb75e75f1/pkg/kubelet/kubelet_pods.go#L975-L980

And I confirmed there are messages by this patch in kubelet log:
```
Dec 10 14:01:20 kind-worker kubelet[559]: I1210 14:01:20.522221     559 kubelet_pods.go:983] "filter terminated pods on status manager" pod="default/out-of-cpu-test-cjxk5" podUID=ec3f822d-59a7-4bef-9b51-0c5e3c6d48b8
```

**Comment by shioshiota** [comment]
I created PR:
https://github.com/kubernetes/kubernetes/pull/106955

**Comment by 249043822** [comment]
Roughly, the scheduler only considers Succeeded or Failed of pod phase as pod terminal state  and thinks that the node has space to place another pod,  but the kubelet considers that both conditions that pod phase is Succeeded or Failed and pod is not actively terminating ( !kl.podWorkers.IsPodTerminationRequested(p.UID) )  are satisfied, then the pod is terminal.  so this is not consistent, this is also the difference between kubelets in v1.21 and v1.22+.

https://github.com/kubernetes/kubernetes/blob/ba8410ea5f34a7404772c2dee873403d340bc7fe/pkg/kubelet/kubelet_pods.go#L979
v1.22+
https://github.com/kubernetes/kubernetes/blob/588a4569a768306ea91821ff916185d65faa85a8/pkg/kubelet/kubelet_pods.go#L923
v1.21-

/cc @smarterclayton

**Comment by smarterclayton** [comment]
This was a deliberately design choice - a Kubelet must consider the resources of terminating pods, and a pod that is terminating but not terminal AND pods that are force deleted (not visible in the API) can still be consuming resources (for instance, volumes or reserved CPUs). Previously, in this case we would be allowing our consumption to burst over what the node could support (and exclusive resources might fail to allocate at all).

I agree that this exposes a conflict between scheduler and kubelet, but it's still possible for the kubelet to have more resources committed than can be observed by the system (a force deleted pod still consumes resources until it is completely terminated) even without my change, and the system needs to act as a whole to minimize that regardless of these temporary blips.

In practice, most production workloads take non-zero time to terminate after deletion.  The scheduler is eager - it tries to place immediately.  Because scheduler + kubelet form a distributed system where network requests can be arbitrarily delayed the kubelet could reject the incoming request because the scheduler is out of date.  The overall goal is to minimize those sorts of conflicts.

We could make the scheduler less eager, but that wouldn't necessarily fix scenarios where multiple schedulers are running and crucially, wouldn't make us more resilient to burst scenarios.  Reverting the change makes the node less resilient.  If we think about this as a race condition, that helps frame the problem more effectively:

> The kubelet has to address how to deal with pods that *could* run in the near future, but *can't* be started yet because the resources they need are not yet available.

Since deletion is one way, we know that eventually the terminating pods will finish.  However, we can't anticipate how long it will be before those pods can be started.  The admission rejection path today gives the scheduler an opportunity to place a pod on another node in some cases (rejecting the pod leads to a workload controller creating another one which might be able to go to another node).  This isn't guaranteed.

There are a couple of scenarios to consider:

1. Pods that want access to exclusive resources (isolated CPUs, PV slots on cloud systems, unique devices)
2. Nodes with a lot of churn (multiple pods starting and terminating)

In both of these cases we could imagine adding a new category of pod admission - pods that *could* run in the near future, once other terminating pods reach terminal state.  That ideally is an unordered list since Kubelets today do not guarantee that property (if a node restarts and has less memory than previously, the kubelet will reject pods semi-arbitrarily).  When checking for admission, we'd allow a pod to pass hard admission if the pod fits within that boundary (active + terminating), but we wouldn't allocate any resources for that pod or start it until those resources become free.

This would require some changes (potentially significant) to kubelet admission.  We'd need to split kubelet pod admission into a verification phase and a mutation phase (like apiserver) and admission plugins like cpu-manager (and others?) would have to be changed to accomodate.  Pods would get verified (can this pod run at all), then accepted (we think we can satisfy this request in the near future), then when we reach soft admission in the pod sync loop is when resources should be allocated (we can start this pod right now).  If an allocation fails, we'd have to be careful to clean it up (a set of partial allocations on stopped pods would be bad).  Likewise, if a pod is in soft admission too long we might simply want to terminate it (although we'd need to think about whether allowing it in creates a perception that it's safe to just wait for the end user).

In brief, that's:

* Pod hard admission has to block impossible requests (requesting 16 cores when we have 8 available)
* Pod hard admission has to allow some future possible requests (requesting 2 cores when we have 0 free but 2 will be released when pod p2 terminates) by taking into account terminating pods + active pods (if we have 8 available cores, 6 running, 0 waiting, and 2 terminating, we can accept 2 cores).  Pod hard admission MUST not accept more than 8 total cores.
* Pod soft admission has to be responsible for deciding whether an admitted but not yet started up can start (wait until there are currently 2 free cpus, then that pod can start)
* All "allocation" of resources should happen in soft admission, and it allocation fails the allocated resources need to be cleaned up inside the admission plugins within some time T 
    - this is a potentially racy situation of course, so it's probably better if admission plugins get passed the set of running pods by the cleanup loop vs doing it themselves (needs follow up)

@derekwaynecarr @ehashman @fromanirh

**Comment by fromanirh** [comment]
@klueska ^^^

**Comment by shioshiota** [comment]
I understood the design idea that kubelet can handle the resources of a terminated pod. And I think it is an excellent design that we introduce hard admission and soft admission in the future. It will solve not only this issue but also more generic problems.

By the way, kubelet 1.22 implementation has this unstable behavior and it doesn't exist before 1.22.
At least, the clusters I manage are affected by this behavior very seriously because the clusters have many batch jobs and many pods start and complete in a short time on the same node.
We have addressed this problem by deploying an in-house patched kubelet(#106955) to our clusters.

I would like to propose introducing a workaround (like the PR #106955 or others) until kubelet handle this race condition well.
I believe that it may help the users who update k8s from 1.21 to 1.22.

**Comment by smarterclayton** [comment]
The proposed workaround regresses the fix that broke other clusters.  It would be better (and generally where we need to get to anyway) to fix the actual problem, and I agree it is important to prioritize to minimize the impact you are seeing.

It doesn't exist before 1.22 because before 1.22 the Kubelet both didn't consider terminating resources in all cases (was broken) AND had other race conditions.  A consequence of the changes to fix that leads to the behavior you are seeing, which is definitely not desirable, but removes a significant amount of uncertainty around whether kubelets could even run the pods it accepted.

**Comment by smarterclayton** [comment]
Description of an e2e test that would have caught this issue (that we should add)

1. Create a large number of pods that run for a known short amount of time that consume 1 or more pod resources (1-5s)
2. When the pod stops, create another pod such that `sum_resources(active + terminating pods) > max_resources_on_node`
3. Verify that no pod fails admission

I suspect but haven't checked that our e2e checks that the pod fails admission (or such a test didn't exist and was added in that state along with these changes), and so the changes in 1.22 locked in the "wrong" user experience.

**Comment by alculquicondor** [comment]
This fits under the category of distributed-systems challenges.
In k8s, using plain Pods is an anti-pattern. When using jobs, you should allow for retries. Arguably, we need more control of what scenarios should be excluded for retries. I plan to revisit this issue #17244

That said, this design change could affect existing workloads when clusters upgrade to 1.22. So we should, at the very least, warn users in the release notes. Perhaps here? https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#known-issues

**Comment by alculquicondor** [comment]
Back to mitigations...

As an alternative to soft-admission: would it be possible that we set a new Pod condition that would allow the scheduler to know whether a pod is fully terminated?

Of course, forced pod deletions would still be a problem.

**Comment by smarterclayton** [comment]
> As an alternative to soft-admission: would it be possible that we set a new Pod condition that would allow the scheduler to know whether a pod is fully terminated?

You mean, so that the scheduler could switch to waiting for pods to be terminated to schedule new pods?  If yes, with changes in 1.22 this should be possible by looking at the pod status (part of the fixes to pod lifecycle were being much more concrete about when the kubelet is available).  When pod is marked as terminating (deletionTimestamp set) AND all containers to have a terminated state in their status, the kubelet is done.  Small possibility we still have a bug there, but we're better at it than we were before.

One downside to such a mitigation is that it would probably increase perceptible latency for rolling updates for apps, possibly in a way that has more dramatic user visible impact.  And long term, kubelet admission is fairly incorrect as is - there are a few places where force deleted pods are not accounted for in admission (https://github.com/kubernetes/kubernetes/issues/104824) that need to be fixed soon.

> Other mitigations

Talked through a few but those that didn't involve the scheduler were strictly worse than fixing kubelet admission.  Most were hacks to kubelet admission that did various subsets of the real fix but then exposed more failures.

> Release notes

Yes, i'll draft something.

**Comment by alculquicondor** [comment]
At this point we have to consider involving scheduler. Not saying that the change is trivial, but probably worth exploring (cc @ahg-g @Huang-Wei ). However, there could be a problem with kubelet versions skew (kubelets that are too old don't populate the container states). What would be the oldest kube-scheduler version we could patch? If I recall the version skew policy correctly, it's just 1.24?

**Comment by smarterclayton** [comment]
Release notes in https://github.com/kubernetes/kubernetes/pull/107348

**Comment by ahg-g** [comment]
There are more issues being opened related to this change, here is another one (Argo workflows): https://github.com/kubernetes/kubernetes/issues/107679

In retrospect, the kubelet change should've been coordinated with sig-scheduling. 

I don't think we can properly address the force pod deletion case, there is no state on the api-server, so nothing the scheduler can do about it.

@smarterclayton what if Kubelet updates pod status to failed/succeeded only when it actually considered the pod terminated and its resource available?

**Comment by Huang-Wei** [comment]
The "OutOf{CPU|Memory}" can also happen when multi schedulers are running - they may race scheduling their pod to the same pod competing for a limited resource. From the user's perspective, the bottom line is the pod can be scheduled elsewhere eventually; instead of staying in this OutOf{CPU|Memory} state forever, unless the landing node's res is released. I'm wondering if we can define an `UnBinding` API to mitigate the problem:

- kubelet issues an `UnBinding` API if it observes a pod stays at OutOf{CPU|Memory} state for a long time
- API server go strips the pod's `.spec.nodeName`
- scheduler act on the event properly (remove the pod from the internal cache and thus is able to reschedule it)

**Comment by ahg-g** [comment]
> The "OutOf{CPU|Memory}" can also happen when multi schedulers are running 

The multi-scheduler problem should probably be addressed by splitting the nodes between schedulers. 

> I'm wondering if we can define an UnBinding API to mitigate the problem

pod lifecycle is already complex, this will only add to its complexity and makes it harder to reason about pod states/phases.

**Comment by smarterclayton** [comment]
@ahg-g 

> what if Kubelet updates pod status to failed/succeeded only when it actually considered the pod terminated and its resources available for other pods?

Possible, although I believe this would be a massive regression in pod end to end latency, and potentially be two orders of magnitude more impactful to end users than the change we instituted here, since pod termination can take up to 30s to clean up resources due to various contention in the kubelet.  It would have a pretty strong impact on jobs, batch, knative, and rolling restart workloads that I believe would show up to almost all users, so I'd be pretty cautious about it without a lot of testing (not that we shouldn't consider it).

One of the real challenges here is that we aren't strict enough in our SLI measurement and SLO for full kube end to end to really be able to even say what the impact of changes were.  We definitely need to minimize the impact of oversubscription around short overlaps.  We definitely want to accurately represent exclusive internal resources in kubelet with lifecycle concerns.  We  *also* want to dramatically reduce total latency from pod create -> pod start -> pod terminate -> pod succeeded/failed showing up.  I do want to systematically attack this, so maybe we can do it by looking at the end to end angle and putting some constraints on, then testing some options.

**Comment by endocrimes** [comment]
/cc
To follow up from sig-node meeting, next steps:
- double check that this is definitely failing on fully-`terminated` pods (bug), or whether this is the expected behavior of a `terminating` container holding up resources (expected behavior, and a container-aware filter is required in the scheduler)

**Comment by ahg-g** [comment]
Thanks @endocrimes

**Comment by princerachit** [comment]
Hello, I see that the comments above say that this issue started occurring with version 1.22 due to kubelet changes. Since this issue is open I believe it has not been fixed any any subsequent version including 1.23? I am asking this because I can see this [comment refering to a fix](https://github.com/kubernetes/kubernetes/issues/106884#issuecomment-990028254). However, the issue of the comment is itself open.

Thanks. Just for context we are sometimes seeing [OutOfMemory error in our 1.23 cluster](https://github.com/gitpod-io/gitpod/issues/7969)

**Comment by ehashman** [comment]
@princerachit that issue is different, where we saw OutOfCpu on node restart. It has been fixed. No restarts are required to trigger the issue reported here.

**Comment by endocrimes** [comment]
So I finally got some time to look at this today. I can't reproduce this on my systems, even after scheduling 6500 pods concurrently with @shioshiota's repro case. (I wonder if Docker for Mac or something has some slow termination issues?).

I'm trying to piece things together from the provided log file now, but with kubelet logs it can be a little tricky to reverse engineer what went on in racier cases.

@shioshiota Could you also provide the logs from the scheduler? It would be helpful here to see the logs from both views of the world (especially bc I cannot repro)

**Comment by liggitt** [comment]
@smarterclayton am I understanding correctly that the kubelet records a pod is terminal in the API _before_ it considers the resources to be free and available for other pods, and then rejects those other pods in admission because it has insufficient resources?

**Comment by smarterclayton** [comment]
To the first part, the kubelet has always done that.  The second part is that the kubelet rejects new pods when there are still pods in non-terminal state.  If we have evidence that new pods are being rejected for terminal pods that the kubelet has informed the apiserver are terminal, that would indeed be a bug (and possibly one made possible by refactors, or simply exposed because we’re more stringent).

**Comment by smarterclayton** [comment]
https://github.com/kubernetes/kubernetes/issues/106884#issuecomment-999390465

identifies the problem correctly:

IsPodTerminationRequested is too broad - we need a boolean check that is the intersection of that state and “isn’t terminating” - ie only the window between terminating and when containers exit.

Sorry for not catching that comment earlier, https://github.com/kubernetes/kubernetes/issues/106884#issuecomment-1032937873 option one is the case.

**Comment by smarterclayton** [comment]
Note the change has to be done with the pod worker lock which means we need a new method IsPodTerminationRequestedAndNotTerminal - i’ll put something up in the morning

**Comment by smarterclayton** [comment]
The e2e that would prevent this regression should be fairly easy to write:

1. create a pod that will have a fairly lengthy termination period due to volume tear down and runs for one second before exiting with code 0, and would constrain the node (perhaps due to a unique resource or simply cpu)
2. create a second pod immediately after the first scheduled that should succeed once the first pod is terminal

I believe the node_e2e don’t catch this because they don’t really use volumes that can have non trivial teardown time or exhaust node resources.  You need non trivial time between container stop and volume cleanup to trigger this reliably, or very high volume.

**Comment by smarterclayton** [comment]
Second note (from bobby in slack) - this only fixes core kubelet resources that don’t require uniqueness (cpu, pods) - resources that are allocated in admission are still broken, if they need a post termination allocation to run.  CPU and device i think check the current state on allocation, but it would be easy to get that wrong and needs to be checked.

**Comment by bobbypage** [comment]
I was able to repro this pretty consistently.

Here is a sample workload:

```
apiVersion: batch/v1
kind: Job
metadata:
  name: busybox
spec:
  completions: 200
  parallelism: 50
  template:
    spec:
      containers:
      - name: busybox
        image: busybox
        resources:
          requests:
            cpu: "1000m"
        command:
        - /bin/sh
        - -c
        - |
          echo "starting!"
          sleep 5s
          _term() {
            echo "trapped sigterm"
            sleep 99999s
          }
          trap _term TERM
      restartPolicy: Never
  backoffLimit: 4
```

(I used one CPU per container, the kind node is running on my workstation which has 12 CPUs)

I created a kind 1.22 cluster with a single node (here is the config: https://gist.github.com/7f13aac0821cee1835f784ae521ce35d)

```
$ kind create cluster --config ${HOME}/dev_scripts/kind-config-e2e.yaml --image kindest/node:v1.22.0@sha256:b8bda84bb3a190e6e028b1760d277454a72267a5454b57db34437c34a588d047
```


Here is status of some of the pods after creating the job:

```
$ kubectl get pods
NAME               READY   STATUS      RESTARTS   AGE
busybox--1-4kgtt   0/1     OutOfcpu    0          7m9s
busybox--1-59ggs   0/1     OutOfcpu    0          7m6s
busybox--1-59k8f   0/1     Completed   0          7m18s
busybox--1-5b2sg   0/1     Completed   0          7m18s
busybox--1-782hq   0/1     Completed   0          7m18s
busybox--1-bdbfk   0/1     Completed   0          7m18s
busybox--1-dh8mz   0/1     Completed   0          7m18s
busybox--1-hvv45   0/1     Completed   0          7m18s
busybox--1-krvdt   0/1     OutOfcpu    0          7m18s
busybox--1-lzpld   0/1     OutOfcpu    0          7m6s
busybox--1-n77d5   0/1     Completed   0          7m18s
busybox--1-vnw6c   0/1     OutOfcpu    0          7m3s
busybox--1-vsm7p   0/1     Completed   0          7m18s
```

**Comment by smarterclayton** [comment]
Do you think you could create a node e2e we can integrate into the fix?

**Comment by bobbypage** [comment]
> Do you think you could create a node e2e we can integrate into the fix?

Yup, I'll work on creating a node e2e for this.

**Comment by wenwu449** [comment]
Looking at "phase" definition at https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodStatus

_Succeeded: All containers in the pod have terminated in success, and will not be restarted. Failed: All containers in the pod have terminated, and at least one container has terminated in failure. The container either exited with non-zero status or was terminated by the system._ 

It is about the state of all **containers**, not about **pod**. Controllers caring about containers rely on this property to be updated ASAP so they can take next action.
Unfortunately that is the only property on pod object can be used by scheduler to determine if the resource are still consumed by the pod. For pod level resource such as "Volume", there is really no good way to tell if the resource has been freed up on the node.

What needed here between kubelet and scheduler is a state of pod resource instead of containers statuses.

**Comment by smarterclayton** [comment]
Given that kubelet shouldn’t be rejecting pods based on resources that are lazily cleaned up (which should reconcile), succeeded and failed should be sufficient.

**Comment by bobbypage** [comment]
I realized that my repro in https://github.com/kubernetes/kubernetes/issues/106884#issuecomment-1050408180 was against v1.22.0 (so I think it was exhibiting issue in https://github.com/kubernetes/kubernetes/pull/104577 rather then the current issue.

However, I was still able to repro it on v1.22.5, though it's slightly more challenging to repro. I created an E2E here - https://github.com/kubernetes/kubernetes/pull/108352

I ran the e2e under kind like [this](https://gist.github.com/b13782876150434807c7f3bd9e56dd3b), and pods were still getting OutOfCPU.

```
$ kubectl get pods -A -o wide
NAMESPACE            NAME                                         READY   STATUS      RESTARTS   AGE    IP             NODE                 NOMINATED NODE   READINESS GATES
job-4362             all-succeed--1-28jc9                         0/1     Completed   0          36s    10.244.2.13    kind-worker2         <none>           <none>
job-4362             all-succeed--1-4l5bb                         0/1     Completed   0          38s    10.244.2.33    kind-worker2         <none>           <none>
job-4362             all-succeed--1-5bkp2                         0/1     Completed   0          38s    10.244.2.4     kind-worker2         <none>           <none>
job-4362             all-succeed--1-5lx96                         0/1     Completed   0          38s    10.244.2.7     kind-worker2         <none>           <none>
job-4362             all-succeed--1-5sj67                         0/1     Completed   0          38s    10.244.2.6     kind-worker2         <none>           <none>
job-4362             all-succeed--1-7hdd9                         0/1     Completed   0          38s    10.244.2.14    kind-worker2         <none>           <none>
job-4362             all-succeed--1-7hjnn                         0/1     Completed   0          36s    10.244.2.9     kind-worker2         <none>           <none>
job-4362             all-succeed--1-99gkw                         0/1     Completed   0          38s    10.244.2.25    kind-worker2         <none>           <none>
job-4362             all-succeed--1-bbqd9                         0/1     Completed   0          38s    10.244.2.3     kind-worker2         <none>           <none>
job-4362             all-succeed--1-bh6cz                         0/1     Completed   0          35s    10.244.2.18    kind-worker2         <none>           <none>
job-4362             all-succeed--1-blbbf                         0/1     Completed   0          38s    10.244.2.8     kind-worker2         <none>           <none>
job-4362             all-succeed--1-cfpw9                         0/1     Completed   0          34s    10.244.2.28    kind-worker2         <none>           <none>
job-4362             all-succeed--1-dmqdn                         0/1     Completed   0          38s    10.244.2.30    kind-worker2         <none>           <none>
job-4362             all-succeed--1-dscnf                         0/1     Completed   0          36s    10.244.2.12    kind-worker2         <none>           <none>
job-4362             all-succeed--1-gcdls                         0/1     Completed   0          38s    10.244.2.21    kind-worker2         <none>           <none>
job-4362             all-succeed--1-h55t6                         0/1     Completed   0          38s    10.244.2.15    kind-worker2         <none>           <none>
job-4362             all-succeed--1-jcp5q                         0/1     Completed   0          36s    10.244.2.10    kind-worker2         <none>           <none>
job-4362             all-succeed--1-jnpxq                         0/1     Completed   0          37s    10.244.2.16    kind-worker2         <none>           <none>
job-4362             all-succeed--1-jx76s                         0/1     Completed   0          38s    10.244.2.24    kind-worker2         <none>           <none>
job-4362             all-succeed--1-lgg25                         0/1     Completed   0          38s    10.244.2.11    kind-worker2         <none>           <none>
job-4362             all-succeed--1-lk6mc                         0/1     Completed   0          34s    10.244.2.20    kind-worker2         <none>           <none>
job-4362             all-succeed--1-lks6n                         0/1     Completed   0          35s    10.244.2.19    kind-worker2         <none>           <none>
job-4362             all-succeed--1-lq8zp                         0/1     Completed   0          38s    10.244.2.5     kind-worker2         <none>           <none>
job-4362             all-succeed--1-sg4kq                         0/1     Completed   0          34s    10.244.2.27    kind-worker2         <none>           <none>
job-4362             all-succeed--1-sqv42                         0/1     OutOfcpu    0          35s    <none>         kind-worker2         <none>           <none>
job-4362             all-succeed--1-t64wb                         0/1     Completed   0          38s    10.244.2.32    kind-worker2         <none>           <none>
job-4362             all-succeed--1-t9jsj                         0/1     Completed   0          38s    10.244.2.31    kind-worker2         <none>           <none>
job-4362             all-succeed--1-tqmwd                         0/1     Completed   0          38s    10.244.2.23    kind-worker2         <none>           <none>
job-4362             all-succeed--1-vvjcg                         0/1     Completed   0          34s    10.244.2.26    kind-worker2         <none>           <none>
job-4362             all-succeed--1-xqpxz                         0/1     Completed   0          38s    10.244.2.22    kind-worker2         <none>           <none>
job-4362             all-succeed--1-z944z                         0/1     Completed   0          36s    10.244.2.17    kind-worker2         <none>           <none>
job-4362             all-succeed--1-zxhnw                         0/1     Completed   0          38s    10.244.2.29    kind-worker2         <none>           <none>
kube-system          coredns-78fcd69978-ctx7m                     1/1     Running     0          89s    10.244.0.3     kind-control-plane   <none>           <none>
kube-system          coredns-78fcd69978-hpj22                     1/1     Running     0          89s    10.244.0.2     kind-control-plane   <none>           <none>
kube-system          etcd-kind-control-plane                      1/1     Running     0          103s   192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kindnet-kmw2j                                1/1     Running     0          74s    192.168.10.3   kind-worker2         <none>           <none>
kube-system          kindnet-mqwqk                                1/1     Running     0          90s    192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kindnet-tft7g                                1/1     Running     0          74s    192.168.10.2   kind-worker          <none>           <none>
kube-system          kube-apiserver-kind-control-plane            1/1     Running     0          103s   192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kube-controller-manager-kind-control-plane   1/1     Running     0          103s   192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kube-proxy-4nktx                             1/1     Running     0          90s    192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kube-proxy-cgkt7                             1/1     Running     0          74s    192.168.10.3   kind-worker2         <none>           <none>
kube-system          kube-proxy-xq8hn                             1/1     Running     0          74s    192.168.10.2   kind-worker          <none>           <none>
kube-system          kube-scheduler-kind-control-plane            1/1     Running     0          104s   192.168.10.4   kind-control-plane   <none>           <none>
local-path-storage   local-path-provisioner-74567d47b4-7qwgd      1/1     Running     0          89s    10.244.0.4     kind-control-plane   <none>           <none>

porterdavid@porterdavid: ~/go/src/k8s.io/kubernetes on issue_106884 [$]
$ kubectl get pods -A -o wide | grep "Out"
job-4362             all-succeed--1-sqv42                         0/1     OutOfcpu    0          37s    <none>         kind-worker2         <none>           <none>
```

**Comment by aojea** [comment]
> I realized that my repro in https://github.com/kubernetes/kubernetes/issues/106884#issuecomment-1050408180 was against v1.22.0 (so I think it was exhibiting issue in https://github.com/kubernetes/kubernetes/pull/104577 rather then the current issue.


@bobbypage I'm publishing kind images with nightlies in https://quay.io/repository/aojea/kindnode?tab=tags , the tag is the kubernetes commit, if you want to check if it happens in master 
```
kind create cluster --name test --image quay.io/aojea/kindnode:e14ace92871
...
kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-02-16T12:38:05Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.0-alpha.3.80+e14ace92871f51", GitCommit:"e14ace92871f513c61a318213a7a2391fbae060b", GitTreeState:"clean", BuildDate:"2022-02-18T01:24:21Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}

```

The 1.23.3 images are already present too https://hub.docker.com/r/kindest/node/tags, you better use kind@main

**Comment by bobbypage** [comment]
Thanks @aojea, I can confirm it also happens on master@e14ace92871 with `quay.io/aojea/kindnode:e14ace92871` and kind@main.

```
porterdavid@porterdavid: ~/go/src/k8s.io/kubernetes on issue_106884 [$]
$ kubectl get nodes -o wide
NAME                 STATUS   ROLES           AGE     VERSION                             INTERNAL-IP    EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION           CONTAINER-RUNTIME
kind-control-plane   Ready    control-plane   2m45s   v1.24.0-alpha.3.80+e14ace92871f51   192.168.10.3   <none>        Ubuntu 21.10   5.14.16-1rodete1-amd64   containerd://1.5.9
kind-worker          Ready    <none>          2m13s   v1.24.0-alpha.3.80+e14ace92871f51   192.168.10.2   <none>        Ubuntu 21.10   5.14.16-1rodete1-amd64   containerd://1.5.9
kind-worker2         Ready    <none>          2m12s   v1.24.0-alpha.3.80+e14ace92871f51   192.168.10.4   <none>        Ubuntu 21.10   5.14.16-1rodete1-amd64   containerd://1.5.9

porterdavid@porterdavid: ~/go/src/k8s.io/kubernetes on issue_106884 [$]
$ kubectl get pods -A -o wide
NAMESPACE            NAME                                         READY   STATUS      RESTARTS   AGE     IP             NODE                 NOMINATED NODE   READINESS GATES
job-3248             all-succeed-6xsm8                            0/1     OutOfcpu    0          100s    <none>         kind-worker2         <none>           <none>
job-3248             all-succeed-l85xf                            0/1     Completed   0          104s    10.244.2.8     kind-worker2         <none>           <none>
job-3248             all-succeed-ldd2q                            0/1     Completed   0          104s    10.244.2.3     kind-worker2         <none>           <none>
job-3248             all-succeed-qzk5t                            0/1     Completed   0          104s    10.244.2.5     kind-worker2         <none>           <none>
job-3248             all-succeed-r5c9g                            0/1     Completed   0          104s    10.244.2.6     kind-worker2         <none>           <none>
job-3248             all-succeed-smmb6                            0/1     Completed   0          104s    10.244.2.7     kind-worker2         <none>           <none>
job-3248             all-succeed-zjfsm                            0/1     Completed   0          104s    10.244.2.4     kind-worker2         <none>           <none>
kube-system          coredns-64897985d-924rf                      1/1     Running     0          2m35s   10.244.0.3     kind-control-plane   <none>           <none>
kube-system          coredns-64897985d-tptcl                      1/1     Running     0          2m35s   10.244.0.2     kind-control-plane   <none>           <none>
kube-system          etcd-kind-control-plane                      1/1     Running     0          2m50s   192.168.10.3   kind-control-plane   <none>           <none>
kube-system          kindnet-qzzn4                                1/1     Running     0          2m20s   192.168.10.4   kind-worker2         <none>           <none>
kube-system          kindnet-t9fsq                                1/1     Running     0          2m21s   192.168.10.2   kind-worker          <none>           <none>
kube-system          kindnet-vz5fz                                1/1     Running     0          2m35s   192.168.10.3   kind-control-plane   <none>           <none>
kube-system          kube-apiserver-kind-control-plane            1/1     Running     0          2m50s   192.168.10.3   kind-control-plane   <none>           <none>
kube-system          kube-controller-manager-kind-control-plane   1/1     Running     0          2m51s   192.168.10.3   kind-control-plane   <none>           <none>
kube-system          kube-proxy-jxnsd                             1/1     Running     0          2m21s   192.168.10.2   kind-worker          <none>           <none>
kube-system          kube-proxy-lmpmv                             1/1     Running     0          2m35s   192.168.10.3   kind-control-plane   <none>           <none>
kube-system          kube-proxy-w9vwq                             1/1     Running     0          2m20s   192.168.10.4   kind-worker2         <none>           <none>
kube-system          kube-scheduler-kind-control-plane            1/1     Running     0          2m50s   192.168.10.3   kind-control-plane   <none>           <none>
local-path-storage   local-path-provisioner-5ddd94ff66-7s8tp      1/1     Running     0          2m35s   10.244.0.4     kind-control-plane   <none>           <none>

porterdavid@porterdavid: ~/go/src/k8s.io/kubernetes on issue_106884 [$]
$ kubectl get pods -A -o wide | grep "Out"
job-3248             all-succeed-6xsm8                            0/1     OutOfcpu    0          107s    <none>         kind-worker2         <none>           <none>
```

[Here](https://gist.github.com/86633b3214955cd6655209e81b57c3d0) are logs from `kind-worker2` on this run.

**Comment by smarterclayton** [comment]
If anyone can reproduce this locally in parallel (I'm still testing some other stuff) can you do it with https://github.com/kubernetes/kubernetes/pull/108356 and see what that log line captures when the out of cpu is caught?  Want to see what pods are in the queue and their phases.

**Comment by smarterclayton** [comment]
I am 99% sure https://github.com/kubernetes/kubernetes/issues/106884#issuecomment-1050729680 does not include force deletions (i don't see the log line  `klog.V(4).InfoS("Gracefully deleting pods from source", "source", source, "pods", klog.KObjs(update.Pods))` showing up)

**Comment by smarterclayton** [comment]
Reproduced with my extra debug:

```
default/busybox-qw6xv(8ce794a6-5371-4f0e-9767-03deb02c08e4) false Pending
default/busybox-s7hx9(e35728fc-90ed-44a7-84a7-f9f5de4d1e13) false Running
default/busybox-vzdsd(79eddd07-7a43-4528-8719-ed230b6865e8) false Pending
default/busybox-89fc9(70f22a93-3087-4c03-8f12-9c4e6e2e62c8) false Running
default/busybox-vhk5b(7bfa54a6-bde6-4b90-8478-5bb7ac82b897) false Pending
default/busybox-7wchx(492435b9-3d31-4a96-a0aa-ad61ace8e545) false Pending
default/busybox-krwjs(04b3cebe-a58f-437e-afd2-649bd04d5745) false Pending
default/busybox-4gxw8(23d3c075-f39a-4af9-a508-5e8d232bfc14) false Pending
default/busybox-nx79w(9f715f04-930c-44e8-802f-c158e57bf386) false Pending
default/busybox-9jg5m(df495b61-e2e0-45ff-994f-9cb782b8ce98) false Pending
default/busybox-rzndp(cced182a-b7b1-4b89-aa0b-8e9a8742deb5) false Running
default/busybox-lbkzj(13b58af0-211a-4ea6-82e7-67d5c1738d1c) false Pending
default/busybox-ddmdv(8b7cb6a0-9efa-41ca-ab2c-3d442a660be3) false Pending
kube-system/coredns-6874cd75d4-tt4pq(979015c6-4275-4914-9d8c-ec6db6acd110) false Running
default/busybox-tq4kc(4a1d20df-1b8d-4683-b893-df32599563f6) false Pending
default/busybox-dztlh(219a82e9-b5fb-4456-82e0-69c4db57c63f) false Pending
default/busybox-dz8b5(8697ba74-e683-4ea6-ab61-5e28b9af41b4) false Succeeded
default/busybox-jtql5(17de0cc4-a254-4bf0-b59d-8b0a260c8ea7) false Running
default/busybox-7f2pr(73b5c3a0-db91-475f-98ac-82b63886e64d) false Pending
default/busybox-rbxdh(43c93813-7311-49e3-b754-c9f86b1819ed) false Pending
default/busybox-x9bhs(8378152c-61c1-4afb-9374-227c2c229a21) false Pending
default/busybox-lmgpk(dc507b86-83be-4e4a-b75e-74675a1b1383) false Pending
default/busybox-42bh2(2c6ee569-23c2-4937-a9ac-ad842526450f) false Running
default/busybox-j59ct(62133300-e95a-4134-a3d4-d345fe072836) false Pending
default/busybox-52xd4(695006aa-e068-49a9-9eb8-9700a8b0e26a) false Pending
default/busybox-9h5jd(52114229-68bd-4506-9bdd-7bb92c7b097e) false Running
default/busybox-bdrj5(cffd6631-fcaa-4583-aa33-dc91b5bd7adb) false Pending
default/busybox-sd9sf(6a78809a-acc8-4021-9c68-ba27270312d0) false Pending
default/busybox-6plnd(be7377b1-a42c-42a1-9588-295c2ddddd6a) false Pending
default/busybox-4dqdj(e33ae860-c2c7-400f-aa93-e23066916fe3) false Pending
default/busybox-kp7m4(51042e7c-65e7-45d0-b591-3a38a8da2e7b) false Pending
default/busybox-5nhn4(66309bed-ed4a-4c36-963f-1315ef988069) false Pending
```
@ 12:02:11.759811

32 pods (1 default in local cluster).  The succeeded pod:

```
I0225 12:02:11.746421  777204 pod_workers.go:630] "Pod is in a terminal phase (success/failed), begin teardown" pod="default/busybox-dz8b5"
I0225 12:02:11.746463  777204 kubelet.go:1719] "syncTerminatingPod enter" pod="default/busybox-dz8b5" podUID=8697ba74-e683-4ea6-ab61-5e28b9af41b4
I0225 12:02:11.746470  777204 kubelet_pods.go:1429] "Generating pod status" pod="default/busybox-dz8b5"
I0225 12:02:11.746487  777204 helpers.go:85] "Already ran container, do nothing" pod="default/busybox-dz8b5" containerName="busybox"
I0225 12:02:11.746502  777204 kubelet_pods.go:1441] "Got phase for pod" pod="default/busybox-dz8b5" oldPhase=Succeeded phase=Succeeded
I0225 12:02:11.746790  777204 status_manager.go:475] "Ignoring same status for pod" pod="default/busybox-dz8b5" status={Phase:Succeeded Conditions:[{T>
I0225 12:02:11.746803  777204 kubelet.go:1749] "Pod terminating with grace period" pod="default/busybox-dz8b5" podUID=8697ba74-e683-4ea6-ab61-5e28b9af>
I0225 12:02:11.759811  777204 predicate.go:114] "DEBUG: failed to fit pod\ndefault/busybox-qw6xv(8ce794a6-5371-4f0e-9767-03deb02c08e4) false Pending\n>
I0225 12:02:11.763755  777204 kuberuntime_manager.go:1035] "getSandboxIDByPodUID got sandbox IDs for pod" podSandboxID=[e037f6f6bb90e194114d011e79722c>
I0225 12:02:11.764384  777204 kubelet.go:1791] "Post-termination container state" pod="default/busybox-dz8b5" podUID=8697ba74-e683-4ea6-ab61-5e28b9af4>
I0225 12:02:11.764395  777204 kubelet.go:1798] "Pod termination stopped all running containers" pod="default/busybox-dz8b5" podUID=8697ba74-e683-4ea6->
I0225 12:02:11.764401  777204 kubelet.go:1800] "syncTerminatingPod exit" pod="default/busybox-dz8b5" podUID=8697ba74-e683-4ea6-ab61-5e28b9af41b4
I0225 12:02:11.764407  777204 pod_workers.go:1014] "Pod terminated all containers successfully" pod="default/busybox-dz8b5" podUID=8697ba74-e683-4e
```

**Comment by bobbypage** [comment]
Rerun again with https://github.com/kubernetes/kubernetes/pull/108356 against master

```
porterdavid@porterdavid: ~/go/src/k8s.io/kubernetes on issue_106884 [$]
$ kubectl get nodes -o wide
NAME                 STATUS   ROLES           AGE    VERSION                              INTERNAL-IP    EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION           CONTAINER-RUNTIME
kind-control-plane   Ready    control-plane   105s   v1.24.0-alpha.3.214+0b0138fac2b15a   192.168.10.4   <none>        Ubuntu 21.10   5.14.16-1rodete1-amd64   containerd://1.5.9
kind-worker          Ready    <none>          85s    v1.24.0-alpha.3.214+0b0138fac2b15a   192.168.10.3   <none>        Ubuntu 21.10   5.14.16-1rodete1-amd64   containerd://1.5.9
kind-worker2         Ready    <none>          72s    v1.24.0-alpha.3.214+0b0138fac2b15a   192.168.10.2   <none>        Ubuntu 21.10   5.14.16-1rodete1-amd64   containerd://1.5.9

porterdavid@porterdavid: ~/go/src/k8s.io/kubernetes on issue_106884 [$]
$ kubectl get pods -A -o wide
NAMESPACE            NAME                                         READY   STATUS      RESTARTS   AGE    IP             NODE                 NOMINATED NODE   READINESS GATES
job-6748             all-succeed-2l857                            0/1     Completed   0          40s    10.244.1.23    kind-worker          <none>           <none>
job-6748             all-succeed-4stg5                            0/1     Completed   0          40s    10.244.1.9     kind-worker          <none>           <none>
job-6748             all-succeed-4w6gj                            0/1     Completed   0          40s    10.244.1.6     kind-worker          <none>           <none>
job-6748             all-succeed-5rn7j                            0/1     Completed   0          38s    10.244.1.14    kind-worker          <none>           <none>
job-6748             all-succeed-6blxz                            0/1     Completed   0          40s    10.244.1.16    kind-worker          <none>           <none>
job-6748             all-succeed-878kh                            0/1     Completed   0          38s    10.244.1.11    kind-worker          <none>           <none>
job-6748             all-succeed-89ffc                            0/1     Completed   0          40s    10.244.1.15    kind-worker          <none>           <none>
job-6748             all-succeed-8hrph                            0/1     Completed   0          39s    10.244.1.13    kind-worker          <none>           <none>
job-6748             all-succeed-d2bxj                            0/1     Completed   0          39s    10.244.1.12    kind-worker          <none>           <none>
job-6748             all-succeed-drf54                            0/1     Completed   0          40s    10.244.1.21    kind-worker          <none>           <none>
job-6748             all-succeed-ggn2m                            0/1     Completed   0          38s    10.244.1.26    kind-worker          <none>           <none>
job-6748             all-succeed-h6rvg                            0/1     Completed   0          40s    10.244.1.3     kind-worker          <none>           <none>
job-6748             all-succeed-k9crh                            0/1     Completed   0          40s    10.244.1.7     kind-worker          <none>           <none>
job-6748             all-succeed-l6ssb                            0/1     Completed   0          38s    10.244.1.18    kind-worker          <none>           <none>
job-6748             all-succeed-lg4sb                            0/1     Completed   0          40s    10.244.1.22    kind-worker          <none>           <none>
job-6748             all-succeed-lrsjm                            0/1     Completed   0          38s    10.244.1.17    kind-worker          <none>           <none>
job-6748             all-succeed-qwb4c                            0/1     Completed   0          38s    10.244.1.8     kind-worker          <none>           <none>
job-6748             all-succeed-rdx8m                            0/1     Completed   0          37s    10.244.1.24    kind-worker          <none>           <none>
job-6748             all-succeed-rgmkt                            0/1     Completed   0          38s    10.244.1.10    kind-worker          <none>           <none>
job-6748             all-succeed-rwkzq                            0/1     Completed   0          40s    10.244.1.4     kind-worker          <none>           <none>
job-6748             all-succeed-rz855                            0/1     Completed   0          40s    10.244.1.19    kind-worker          <none>           <none>
job-6748             all-succeed-t5gpw                            0/1     Completed   0          40s    10.244.1.20    kind-worker          <none>           <none>
job-6748             all-succeed-tx7rh                            0/1     OutOfcpu    0          37s    <none>         kind-worker          <none>           <none>
job-6748             all-succeed-xlslh                            0/1     Completed   0          40s    10.244.1.5     kind-worker          <none>           <none>
kube-system          coredns-64897985d-7jmrd                      1/1     Running     0          93s    10.244.0.3     kind-control-plane   <none>           <none>
kube-system          coredns-64897985d-wnnl8                      1/1     Running     0          93s    10.244.0.2     kind-control-plane   <none>           <none>
kube-system          etcd-kind-control-plane                      1/1     Running     0          106s   192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kindnet-p4kbr                                1/1     Running     0          93s    192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kindnet-rdbdn                                1/1     Running     0          77s    192.168.10.2   kind-worker2         <none>           <none>
kube-system          kindnet-tn6ps                                1/1     Running     0          90s    192.168.10.3   kind-worker          <none>           <none>
kube-system          kube-apiserver-kind-control-plane            1/1     Running     0          106s   192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kube-controller-manager-kind-control-plane   1/1     Running     0          106s   192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kube-proxy-42slk                             1/1     Running     0          77s    192.168.10.2   kind-worker2         <none>           <none>
kube-system          kube-proxy-7wxxn                             1/1     Running     0          90s    192.168.10.3   kind-worker          <none>           <none>
kube-system          kube-proxy-p2zv6                             1/1     Running     0          93s    192.168.10.4   kind-control-plane   <none>           <none>
kube-system          kube-scheduler-kind-control-plane            1/1     Running     0          106s   192.168.10.4   kind-control-plane   <none>           <none>
local-path-storage   local-path-provisioner-5ddd94ff66-zbkj9      1/1     Running     0          93s    10.244.0.4     kind-control-plane   <none>           <none>

porterdavid@porterdavid: ~/go/src/k8s.io/kubernetes on issue_106884 [$]
$ kubectl get pods -A -o wide | grep "Out"
job-6748             all-succeed-tx7rh                            0/1     OutOfcpu    0          39s    <none>         kind-worker          <none>           <none>

```

kind-worker logs - https://gist.github.com/2f019b52f17eb133b50802fb469fd603

**Comment by smarterclayton** [comment]
The bug is the previously theorized (last night, after the first theory didn't pan out) "pods that exit early cleanly are updating terminal phase in apiserver too soon, instead of doing it after terminating".

We shouldn't set the phase until we are 100% certain that no future containers can be started, and that all containers are stopped.  I think the current code can do the first part (we're in syncPod when we detect this, so if syncPod exits no future containers will be created), but the second part may not actually be accurate from a kubelet perspective (have to think about it).

Reticulating splines...

**Comment by shioshiota** [comment]
@smarterclayton 
I reproduced with https://github.com/kubernetes/kubernetes/pull/108356
https://gist.github.com/shioshiota/27ced280497e5c9c078cf33e802db82b
@endocrimes 
And attached scheduler log. I'm sorry to reply too late.

**Comment by smarterclayton** [comment]
The most straightforward fix (read: most likely not to dramatically regress anything other than maybe some shutdown latency) is probably:

> Kubelet shouldn't tell the apiserver about transition to a terminal phase until the kubelet has verified the terminal phase (podWorker.CouldHaveRunningContainers() returns false)

That adds latency, but not that unmanageable.  We probably want to exit syncPod early if we detect the pod is now in a terminal phase too.

I really don't know what to do with eviction - with the proposed fix above, if the kubelet restarts after starting eviction, we'd lose any knowledge that we tried to evict the pod.  We *could* accept the latency in the short term, and then add a new condition `Evicted` that has a clear message, and then if we detected the condition we'd simply treat that as if we recorded an eviction.  That has other advantages really...

**Comment by alculquicondor** [comment]
+1 on explicit Eviction condition (post hotfix)

**Comment by smarterclayton** [comment]
Correctly delaying updates that set a terminal phase from reaching the apiserver is challenging to implement safely without impacting other status updates.  Going to need some time to think about how to work a change in that doesn't risk regressing in another subtle way.

**Comment by xhejtman** [comment]
Hello, tested 1.23.4 and outofcpu bug is still present. It is not only for terminating pods. It happens also in case, you spawn like 400 new pods and not all can be executed so they are pending and some of them end with outofcpu even before running. This causes e.g., nextflow workflow manager so for this tool, 1.22+ is currently unusable as one needs to restart the pipeline or impose only small limit on number of parallel pods. The pod resource requests vary (like 16 cpus/pod or 1 cpu/pod), so you cannot simply reduce the in flight pods to match available resources and nextflow does not contain any other means of resource limit.

**Comment by rphillips** [comment]
@xhejtman The fix will be in 1.23.5

**Comment by FullyScaled** [comment]
We still see similiar issues (but with OutOfMemory instead of OutOfCpu) with k8s 1.23.4. 

@rphillips Would you expect those issues to be addressed with k8s version 1.23.6 as well?

**Comment by AshutoshNirkhe** [comment]
I am seeing 'OutOfCpu' still in v1.26.3

**Comment by alculquicondor** [comment]
Please provide context. Some OutOfCpu errors are legitimate, depending on the circumstances. Ideally, open a new issue and reference the old one.

**Comment by adecchi-2inno** [comment]
Same issue at 1.27.4-gke.900

**Comment by Kavinjsir** [comment]
Hi folks, we are running `1.28.3-gke.1203001` and seems similar error logs remain there for several times. Not sure if it is relative to the thread here, would there be someone bringing hints/insights? Thx!

**Comment by brennenmurray** [comment]
@Kavinjsir I'm having the same issue but in 1.27 I think GKE is having issues in regards to descheduling failed pods from nodes that come offline. All my pods that have this OutOfCpu or OutOfMemory are all spot nodes. Maybe a new taint for the node problem detector in the KCP not picking up on these dead nodes for the scheduler? Fun times

**Comment by gdubicki** [comment]
I have been seeing this in `1.28.7-gke.1026001` too recently.

**Comment by kannon92** [comment]
Not sure why it says I closed it.

/reopen

**Comment by btrajkovski-mms** [comment]
This issue happened to me in `1.29.7-gke.1104000` while deploying 2 workloads in parallel. 282 pods were created and stuck in `OutOfCpu` state.

**Comment by alculquicondor** [comment]
/close
as duplicate of #115325
